dlrover/python/common/event/context.py

```diff
@@ -77,80 +96,115 @@ def add_step_event(self, event: AtorchEvent):
+                self._step_events.popitem(last=False)
 
-            last_key = keys[-1]
-            last_step = self._step_events[last_key]
-            if last_step.end_timestamp is None:
-                logger.error(f"invalid last step: {last_step}")
-                return
-            if (
-                event.step != last_step.step
-                or event.timestamp < last_step.end_timestamp
-            ):
-                logger.error(f"invalid last step: {last_step}, {event}")
-                return
+            if event.type == EventTypeName.BEGIN:
+                if len(keys) > 0:
+                    last_event = self._step_events[keys[-1]]
+                    if last_event.event_state != TrainEventState.TRAIN_EVT_END:
+                        logger.error(
+                            f"invalid ckpt step: {last_event}, {event}"
+                        )
+                        return
+                    if event.timestamp < last_event.end_timestamp:
+                        logger.error(
+                            f"invalid ckpt step time: {last_event}, {event}"
+                        )
+                        return
+
+                step_event = AtorchStepEvent(
+                    event.name, TrainEventState.TRAIN_EVT_BEGIN, event.step
+                )
+                step_event.begin_timestamp = event.timestamp
+                step_event.localtime = int(datetime.now().timestamp())
+                self._step_events[event.timestamp] = step_event
+
+            elif event.type == EventTypeName.END:
+                if not len(keys):
+                    logger.error(f"invalid ckpt step without BEGIN: {event}")
+                    return
+                last_key = keys[-1]
+                step_event = self._step_events[last_key]
+                if (
+                    step_event.step != event.step
+                    or step_event.event_state
+                    != TrainEventState.TRAIN_EVT_BEGIN
+                ):
+                    logger.error(f"invalid ckpt step: {step_event}, {event}")
+                    return
+                if step_event.begin_timestamp > event.timestamp:
+                    logger.error(
+                        f"invalid ckpt step time: {step_event}, {event}"
+                    )
+                    return
 
-            if event.type is EventTypeName.BEGIN:
-                last_step.start_ckpt(event.timestamp)
-            elif event.type is EventTypeName.END:
-                last_step.finish_ckpt(event.timestamp)
-                last_step.localtime = int(datetime.now().timestamp())
+                step_event.end_timestamp = event.timestamp
+                step_event.step_time = (
+                    step_event.end_timestamp - step_event.begin_timestamp
+                )
+                step_event.step = event.step
+                step_event.event_state = TrainEventState.TRAIN_EVT_END
+                step_event.localtime = int(datetime.now().timestamp())
```

Comment:
int(datetime.now().timestamp()) -> int(time.time())
Since we are already using Unix timestamps here, there's no need to add an extra conversion through datetime.